zerver/lib/import_realm.py

```diff
@@ -2041,61 +2052,99 @@ def get_incoming_message_ids(import_dir: Path, sort_by_date: bool) -> list[int]:
-        bulk_import_model(data, Message)
-
-        # Due to the structure of these message chunks, we're
-        # guaranteed to have already imported all the Message objects
-        # for this batch of UserMessage objects.
-        re_map_foreign_keys(data, "zerver_usermessage", "message", related_table="message")
-        re_map_foreign_keys(
-            data, "zerver_usermessage", "user_profile", related_table="user_profile"
-        )
-        fix_bitfield_keys(data, "zerver_usermessage", "flags")
 
-        bulk_import_user_message_data(data, dump_file_id)
-        dump_file_id += 1
+def _process_message_file(
+    dump_file_id: int, realm: Realm, sender_map: dict[int, Record], import_dir: Path
+) -> None:
+    """Process a single message file (used for sequential import)."""
+    message_filename = os.path.join(import_dir, f"messages-{dump_file_id:06}.json")
+
+    with open(message_filename, "rb") as f:
+        data = orjson.loads(f.read())
+
+    re_map_foreign_keys(data, "zerver_message", "sender", related_table="user_profile")
+    re_map_foreign_keys(data, "zerver_message", "recipient", related_table="recipient")
+    re_map_foreign_keys(data, "zerver_message", "sending_client", related_table="client")
+    fix_datetime_fields(data, "zerver_message")
+    fix_upload_links(data, "zerver_message")
+
+    message_id_map = ID_MAP["message"]
+    for row in data["zerver_message"]:
+        del row["realm"]
+        row["realm_id"] = realm.id
+        row["id"] = message_id_map[row["id"]]
+
+    for row in data["zerver_usermessage"]:
+        assert row["message"] in message_id_map
+
+    fix_message_rendered_content(
+        realm=realm,
+        sender_map=sender_map,
+        messages=data["zerver_message"],
+    )
+    logging.info("Successfully rendered Markdown for message batch")
+
+    fix_message_edit_history(realm=realm, sender_map=sender_map, messages=data["zerver_message"])
+    bulk_import_model(data, Message)
+
+    re_map_foreign_keys(data, "zerver_usermessage", "message", related_table="message")
+    re_map_foreign_keys(data, "zerver_usermessage", "user_profile", related_table="user_profile")
+    fix_bitfield_keys(data, "zerver_usermessage", "flags")
+
+    bulk_import_user_message_data(data, dump_file_id)
+
+
+def _import_message_file_worker(data: tuple[int, int, dict[int, Record], Path]) -> None:
+    """Worker function for parallel message import."""
+    dump_file_id, realm_id, sender_map, import_dir = data
+
+    # Load ID_MAP in worker
+    id_map_path = os.path.join(import_dir, "id_map.json")
+    with open(id_map_path, "rb") as f:
+        loaded_id_map = orjson.loads(f.read())
+
+    # Convert string keys back to integers
+    for table_name in loaded_id_map:
+        if isinstance(loaded_id_map[table_name], dict):
+            loaded_id_map[table_name] = {int(k): v for k, v in loaded_id_map[table_name].items()}
+
+    ID_MAP.update(loaded_id_map)
+
+    realm = Realm.objects.get(id=realm_id)
```

Comment:
Doing this in the initializer means that we won't repeat this work for every batch.